src/tray/windows/SDL_tray.c

```diff
@@ -0,0 +1,579 @@
+{
+    if (!tray->menu) {
+        return NULL;
+    }
+
+    return find_entry_in_menu(tray->menu, id);
+}
+
+LRESULT CALLBACK TrayWindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam) {
+    SDL_Tray *tray = (SDL_Tray *) GetWindowLongPtr(hwnd, GWLP_USERDATA);
+    SDL_TrayEntry *entry = NULL;
+
+    if (!tray) {
+        return DefWindowProc(hwnd, uMsg, wParam, lParam);
+    }
+
+    switch (uMsg) {
+        case WM_TRAYICON:
+            if (lParam == WM_RBUTTONUP || lParam == WM_LBUTTONUP) {
+                POINT pt;
+                GetCursorPos(&pt);
+                SetForegroundWindow(hwnd);
+                TrackPopupMenu(tray->menu->hMenu, TPM_BOTTOMALIGN | TPM_RIGHTALIGN, pt.x, pt.y, 0, hwnd, NULL);
+            }
+            break;
+
+        case WM_COMMAND:
+            entry = find_entry_with_id(tray, LOWORD(wParam));
+
+            if (entry && (entry->flags & SDL_TRAYENTRY_CHECKBOX)) {
+                SDL_SetTrayEntryChecked(entry, !SDL_GetTrayEntryChecked(entry));
+            }
+
+            if (entry && entry->callback) {
+                entry->callback(entry->userdata, entry);
+            }
+            break;
+
+        default:
+            return DefWindowProc(hwnd, uMsg, wParam, lParam);
+    }
+    return 0;
+}
+
+static void DestroySDLMenu(SDL_TrayMenu *menu)
+{
+    for (size_t i = 0; i < menu->nEntries; i++) {
+        if (menu->entries[i] && menu->entries[i]->submenu) {
+            DestroySDLMenu(menu->entries[i]->submenu);
+        }
+        SDL_free(menu->entries[i]);
+    }
+    SDL_free(menu->entries);
+    DestroyMenu(menu->hMenu);
+    SDL_free(menu);
+}
+
+static wchar_t *convert_label(const char *in)
+{
+    const char *c;
+    char *c2;
+    int len = 0;
+
+    for (c = in; *c; c++) {
+        len += (*c == '&') ? 2 : 1;
+    }
+
+    char *escaped = SDL_malloc(SDL_strlen(in) + len + 1);
+
+    if (!escaped) {
+        return NULL;
+    }
+
+    for (c = in, c2 = escaped; *c;) {
+        if (*c == '&') {
+            *c2++ = *c;
+        }
+
+        *c2++ = *c++;
+    }
+
+    *c2 = '\0';
+
+    int len_w = MultiByteToWideChar(CP_UTF8, 0, escaped, len, NULL, 0);
+    wchar_t *out = (wchar_t *)SDL_malloc(len_w * sizeof(wchar_t));
+
+    if (!out) {
+        SDL_free(escaped);
+        return NULL;
+    }
+
+    MultiByteToWideChar(CP_UTF8, 0, escaped, -1, out, len_w);
+
+    SDL_free(escaped);
+
+    return out;
+}
+
+SDL_Tray *SDL_CreateTray(SDL_Surface *icon, const char *tooltip)
+{
+    SDL_Tray *tray = SDL_malloc(sizeof(SDL_Tray));
+
+    if (!tray) {
+        return NULL;
+    }
+
+    tray->hwnd = NULL;
+    tray->menu = NULL;
+
+    wchar_t classname[32] = { L'\0' };
+    SDL_swprintf(classname, 32, L"SDLTray%d", (unsigned int) get_next_id());
+
+    HINSTANCE hInstance = GetModuleHandle(NULL);
+    WNDCLASSW wc;
+    ZeroMemory(&wc, sizeof(WNDCLASS));
+    wc.lpfnWndProc = TrayWindowProc;
+    wc.hInstance = hInstance;
+    wc.lpszClassName = classname;
+
+    RegisterClassW(&wc);
```

Comment:
It is most likely not necessary, I'm just unfamiliar with Win32 :)

I originally tried to re-use the existing SDL window class to avoid duplicating code, but I found that this created an unnecessary dependency on SDL_Window, so I later moved the code over. It's probably still visible in the structure. I'll try to fix that now; please let me know if there are any other mistakes like that.